crates/egui/src/memory.rs

```diff
@@ -322,31 +355,121 @@ impl Focus {
                 self.give_to_next = false;
-                self.pressed_tab = false;
-                self.pressed_shift_tab = false;
+                self.reset_focus();
             }
         }
 
+        // The rect is updated at the end of the frame.
+        self.focus_widgets_cache
+            .entry(id)
+            .or_insert(Rect::EVERYTHING);
+
         if self.give_to_next && !self.had_focus_last_frame(id) {
             self.id = Some(id);
             self.give_to_next = false;
         } else if self.id == Some(id) {
-            if self.pressed_tab && !self.is_focus_locked {
+            if matches!(self.focus_direction, FocusDirection::Next) && !self.is_focus_locked {
                 self.id = None;
                 self.give_to_next = true;
-                self.pressed_tab = false;
-            } else if self.pressed_shift_tab && !self.is_focus_locked {
+                self.reset_focus();
+            } else if matches!(self.focus_direction, FocusDirection::Previous)
+                && !self.is_focus_locked
+            {
                 self.id_next_frame = self.last_interested; // frame-delay so gained_focus works
-                self.pressed_shift_tab = false;
+                self.reset_focus();
             }
-        } else if self.pressed_tab && self.id.is_none() && !self.give_to_next {
+        } else if matches!(self.focus_direction, FocusDirection::Next)
+            && self.id.is_none()
+            && !self.give_to_next
+        {
             // nothing has focus and the user pressed tab - give focus to the first widgets that wants it:
             self.id = Some(id);
-            self.pressed_tab = false;
+            self.reset_focus();
         }
 
         self.last_interested = Some(id);
     }
+
+    pub fn reset_focus(&mut self) {
+        self.focus_direction = FocusDirection::None;
+    }
+
+    pub fn find_widget_in_direction(&mut self, new_rects: &IdMap<Rect>) -> Option<Id> {
+        let Some(focus_id) = self.id else {return None;};
+
+        // Update cache with new rects
+        self.focus_widgets_cache.retain(|id, old_rect| {
+            if let Some(new_rect) = new_rects.get(id) {
+                *old_rect = *new_rect;
+                true // Keep the item
+            } else {
+                false // Remove the item
+            }
+        });
+
+        let current_focus_widget_rect = self.focus_widgets_cache.get(&focus_id).unwrap();
+        let current_widget_pos = current_focus_widget_rect.left_center();
+
+        let mut focus_candidates: Vec<(&Id, f32, f32)> =
+            Vec::with_capacity(self.focus_widgets_cache.len());
+
+        for (widget_id, widget_rect) in &mut self.focus_widgets_cache {
+            if Some(*widget_id) == self.id {
+                continue;
+            }
+
+            let candidate_widget_pos = widget_rect.left_center();
+
+            let current_to_candidate = candidate_widget_pos - current_widget_pos;
```

Comment:
This could be improved. Instead of taking an arbitrary point (`left_center`) in the rect, we should consider the full rects.

For instance, these right-aligned rects should be considered to be perfectly above each other:

```
[ first widget ]
     [ another ]
        [ last ]
```

We can accomplish this with a function that checks the x-ranges and y-ranges and if they overlap, returns 0 for that axis.

```rs
/// * negative if `a` is strictly left of `b`
/// * positive if `a` is strictly right of `b`
/// * zero if the ranges overlap
fn range_diff(a: Rangef, b: Rangef) -> f32 {
    if a.max < b.min {
        a.max - b.min
    } else if b.max < a.min {
        a.min - b.max
    } else {
        0.0 // overlaps
    }
}

…

let current_to_candate = vec2(
    range_diff(selected_rect.x_range(), candidate_rect.x_range(),
    range_diff(selected_rect.y_range(), candidate_rect.y_range(),
);
```

I'm not a 100% this is an improvement, but I believe so.